rholang/src/main/scala/coop/rchain/rholang/interpreter/RholangCLI.scala

```diff
@@ -120,26 +122,30 @@ object RholangCLI {
 
   def buildNormalizedTerm(source: Reader): Either[Throwable, Par] =
     try {
-      val term = buildAST(source)
-      val inputs =
-        ProcVisitInputs(VectorPar(), DebruijnIndexMap[VarSort](), DebruijnLevelMap[VarSort]())
-      val normalizedTerm = normalizeTerm(term, inputs).leftMap(s => new Exception(s))
-      normalizedTerm.flatMap { (nt: ProcVisitOutputs) =>
-        ParSortMatcher
-          .sortMatch(Some(nt.par))
-          .term
-          .fold[Either[Throwable, Par]](
-            Left[Throwable, Par](new Exception("ParSortMatcher failed")))(p => Right(p))
-      }
+      for {
+        term <- buildAST(source)
+        inputs = ProcVisitInputs(VectorPar(),
+                                 DebruijnIndexMap[VarSort](),
+                                 DebruijnLevelMap[VarSort]())
+        outputs <- normalizeTerm(term, inputs).leftMap(s => new Exception(s))
+        par <- Either.fromOption(ParSortMatcher.sortMatch(Some(outputs.par)).term,
+                                 new Exception("ParSortMatcher failed"))
+      } yield par
     } catch {
-      case ex: Throwable => Left(ex)
+      case th: Throwable => Left(th)
     }
 
-  private def buildAST(source: Reader): Proc = {
-    val lxr = lexer(source)
-    val ast = parser(lxr)
-    ast.pProc()
-  }
+  private def buildAST(source: Reader): Either[Throwable, Proc] =
+    Either
+      .catchNonFatal {
+        val lxr = lexer(source)
+        val ast = parser(lxr)
+        ast.pProc()
+      }
+      .leftMap {
+        case ex: Exception if ex.getMessage.toLowerCase.contains("syntax") =>
```

Comment:
Is the parser really not throwing a distinguishable error?

If so, we should try and fix that.

Reply:
Yup, that's the snippet from generated `parser.java` file:
```
public void unrecovered_syntax_error(java_cup.runtime.Symbol cur_token) throws java.lang.Exception
{
  throw new Exception("Unrecoverable Syntax Error");
}
```